rust/agents/relayer/src/msg/metadata/multisig.rs

```diff
@@ -0,0 +1,195 @@
+        else {
+            info!(
+                ?validators, threshold, highest_known_nonce,
+                "Could not fetch metadata: Unable to reach quorum"
+            );
+            return Ok(None);
+        };
+
+        // At this point we have a signed checkpoint with a quorum of validator
+        // signatures. But it may be a fraudulent checkpoint that doesn't
+        // match the canonical root at the checkpoint's index.
+        debug!(?checkpoint, "Found checkpoint with quorum");
+
+        let proof = self
+            .prover_sync
+            .read()
+            .await
+            .get_proof(message.nonce, checkpoint.checkpoint.index)
+            .context(CTX)?;
+
+        if checkpoint.checkpoint.root == proof.root() {
+            debug!(
+                ?validators,
+                threshold,
+                ?checkpoint,
+                ?proof,
+                "Fetched metadata"
+            );
+            let metadata = self.format_metadata(&validators, threshold, &checkpoint, &proof);
+            Ok(Some(metadata))
+        } else {
+            info!(
+                ?checkpoint,
+                canonical_root = ?proof.root(),
+                "Could not fetch metadata: Signed checkpoint does not match canonical root"
+            );
+            Ok(None)
+        }
+    }
+}
+
+impl MultisigIsmMetadataBuilder {
+    /// Returns the metadata needed by the contract's verify function
+    fn format_metadata(
+        &self,
+        validators: &[H256],
+        threshold: u8,
+        checkpoint: &MultisigSignedCheckpoint,
+        proof: &Proof,
+    ) -> Vec<u8> {
+        assert_eq!(threshold as usize, checkpoint.signatures.len());
+        let root_bytes = checkpoint.checkpoint.root.to_fixed_bytes().into();
+        let index_bytes = checkpoint.checkpoint.index.to_be_bytes().into();
+        let proof_tokens: Vec<Token> = proof
+            .path
+            .iter()
+            .map(|x| Token::FixedBytes(x.to_fixed_bytes().into()))
+            .collect();
+        let mailbox_and_proof_bytes = ethers::abi::encode(&[
+            Token::FixedBytes(
+                checkpoint
+                    .checkpoint
+                    .mailbox_address
+                    .to_fixed_bytes()
+                    .into(),
+            ),
+            Token::FixedArray(proof_tokens),
+        ]);
+
+        // The ethers encoder likes to zero-pad non word-aligned byte arrays.
+        // Thus, we pack the signatures, which are not word-aligned, ourselves.
+        let signature_vecs: Vec<Vec<u8>> = order_signatures(validators, &checkpoint.signatures);
+        let signature_bytes = signature_vecs.concat();
+
+        let validator_tokens: Vec<Token> = validators
+            .iter()
+            .map(|x| Token::FixedBytes(x.to_fixed_bytes().into()))
+            .collect();
+        let validator_bytes = ethers::abi::encode(&[Token::FixedArray(validator_tokens)]);
+        if !self.legacy {
```

Comment:
nit: invert these?

Reply:
I prefer it in this order because the "normal" case is first... but I do see the argument for either ordering and don't care that much.

Reply:
Yeah that was my rationale but I don't feel strongly